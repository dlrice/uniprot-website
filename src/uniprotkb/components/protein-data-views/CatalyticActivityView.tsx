import React, { Fragment, useState, useCallback } from 'react';
import '@swissprot/rhea-reaction-visualizer';
import { useModal, ModalBackdrop, Window } from 'franklin-sites';
import UniProtKBEvidenceTag from './UniProtKBEvidenceTag';
import {
  CatalyticActivityComment,
  PhysiologicalReactionDirection,
  PhysiologicalReaction,
} from '../../types/commentTypes';
import './styles/catalytic-activity-view.scss';

// example accession to view this component: P31937

export const getRheaId = (referenceId: string) => {
  const re = /^RHEA:(\d+)$/i;
  const match = referenceId.match(re);
  return match && parseInt(match[1], 10);
};

export const isRheaReactionReference = ({
  database,
  id,
}: {
  database: string;
  id: string;
}) => database === 'Rhea' && !!getRheaId(id);

type RheaReactionVisualizerProps = {
  rheaId: number;
  show: boolean;
};

export const RheaReactionVisualizer: React.FC<RheaReactionVisualizerProps> = ({
  rheaId,
  show: initialShow,
}) => {
  const [show, setShow] = useState(initialShow);
  const [zoomImageUrl, setZoomImageUrl] = useState('');
  const { displayModal, setDisplayModal, Modal } = useModal(
    ModalBackdrop,
    Window
  );
  const callback = useCallback((node): void => {
    if (node) {
      node.addEventListener(
        'zoomClicked',
        ({ detail }: { detail: { chebi: string; imgURL: string } }) => {
          // eslint-disable-next-line no-console
          console.log('zoomClicked:', detail);
          setZoomImageUrl(detail.imgURL);
          setDisplayModal(true);
        }
      );
    }
  }, []);

  return (
    <Fragment>
      <button
        type="button"
        className="button tertiary rhea-reaction-visualizer__button"
        onClick={() => setShow(!show)}
      >
        {`${show ? 'Hide' : 'View'} Rhea reaction`}
      </button>
      {show && (
        <Fragment>
          <div className="rhea-reaction-visualizer__component">
            <rhea-reaction rheaid={rheaId} zoom showids={true} ref={callback} />
          </div>
          {displayModal && zoomImageUrl && (
            <Modal
              handleExitModal={() => setDisplayModal(false)}
              width="50vw"
              height="50vh"
              withFooterCloseButton={true}
            >
              <img src={zoomImageUrl} />
            </Modal>
          )}
        </Fragment>
      )}
    </Fragment>
  );
};

export const physiologicalReactionDirectionToString = new Map<
  PhysiologicalReactionDirection,
  string
>([
  [PhysiologicalReactionDirection.LeftToRight, 'forward'],
  [PhysiologicalReactionDirection.RightToLeft, 'backward'],
]);

export type ReactionDirectionProps = {
  physiologicalReactions: PhysiologicalReaction[];
};

export const ReactionDirection: React.FC<ReactionDirectionProps> = ({
  physiologicalReactions,
}) => {
  /*
  Possible output:
    1. This reaction proceeds in the backward direction <Evidence>
    2. This reaction proceeds in the forward direction <Evidence>
    3. This reaction proceeds in the forward <Evidence> and the backward <Evidence> directions.
  */
  if (!physiologicalReactions || physiologicalReactions.length === 0) {
    return null;
  }
  if (physiologicalReactions.length > 2) {
    // eslint-disable-next-line no-console
    console.error(
      'More than two physiological reactions encountered when rendering catalytic activity'
    );
    return null;
  }
  return (
    <Fragment>
      {`This reaction proceeds in `}
      {physiologicalReactions
        // Ensure that left-to-right/forward comes before right-to-left/backward
        .sort((a, b) => a.directionType.localeCompare(b.directionType))
        .map(({ reactionCrossReference, directionType, evidences }, index) => (
          <Fragment key={reactionCrossReference.id}>
            {index > 0 && ' and '}
            {`the `}
            <span data-testid="direction-text">
              {physiologicalReactionDirectionToString.get(directionType)}
            </span>
            {physiologicalReactions.length === 1 && ' direction '}
            <UniProtKBEvidenceTag evidences={evidences} />
            {physiologicalReactions.length === 2 &&
              index === 1 &&
              ' directions '}
          </Fragment>
        ))}
    </Fragment>
  );
};

type CatalyticActivityProps = {
  comments?: CatalyticActivityComment[];
  title?: string;
};

const CatalyticActivityView: React.FC<CatalyticActivityProps> = ({
  comments,
  title,
}) => {
  if (!comments || comments.length <= 0) {
    return null;
  }
  let firstRheaId: number | null = null;
  return (
    <Fragment>
      {title && <h3>{title}</h3>}
      {comments.map(({ reaction, physiologicalReactions }) => {
        if (!reaction) {
          return null;
        }
        // Using only the first rhea reaction reference because FW has assured us that
        // there will be either 0 or 1 types of this reference (ie never > 1)

        const rheaReactionReference =
          reaction.reactionCrossReferences &&
          reaction.reactionCrossReferences.find(isRheaReactionReference);
        const rheaId =
          rheaReactionReference && getRheaId(rheaReactionReference.id);
        if (rheaId && !firstRheaId) {
          firstRheaId = rheaId;
        }
        return (
          <span className="text-block" key={reaction.ecNumber}>
            <strong>{reaction.ecNumber}</strong>
            {/* Need a link to search for EC in UniProtKB:
             https://www.ebi.ac.uk/panda/jira/browse/TRM-23597 */}
            {` ${reaction.name}`}
            {reaction.evidences && (
              <UniProtKBEvidenceTag evidences={reaction.evidences} />
            )}
            {physiologicalReactions && physiologicalReactions.length && (
              <ReactionDirection
                physiologicalReactions={physiologicalReactions}
              />
            )}
            {!!rheaId && (
              <RheaReactionVisualizer
                rheaId={rheaId}
                show={rheaId === firstRheaId}
              />
            )}
          </span>
        );
      })}
    </Fragment>
  );
};

export default CatalyticActivityView;
