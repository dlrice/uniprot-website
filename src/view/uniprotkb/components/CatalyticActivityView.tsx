import React, { Fragment, useState } from 'react';
import '@swissprot/rhea-reaction-visualizer';
import UniProtEvidenceTag from '../../../components/UniProtEvidenceTag';
import { Evidence } from '../../../model/types/modelTypes';
import Comment from '../../../model/types/Comment';
import './styles/CatalyticActivityView.scss';

export const getRheaId = (referenceId: string) => {
  const re = /^RHEA:(\d+)$/;
  const match = referenceId.match(re);
  return match && match[1];
};

export const isRheaReactReference = ({
  databaseType,
  id,
}: {
  databaseType: string;
  id: string;
}) => databaseType === 'Rhea' && !!getRheaId(id);

type RheaReactionVisualizerProps = {
  rheaId: string;
  show: boolean;
};

export const RheaReactionVisualizer: React.FC<RheaReactionVisualizerProps> = ({
  rheaId,
  show: initialShow,
}) => {
  const [show, setShow] = useState(initialShow);
  return (
    <Fragment>
      <button
        type="button"
        className="button link-button rhea-reaction-visualizer__button"
        onClick={() => setShow(!show)}
      >
        {`${show ? 'Hide' : 'View'} Rhea reaction`}
      </button>
      {show && (
        <div className="rhea-reaction-visualizer__component">
          <rhea-reaction rheaid={rheaId} zoom showids />
        </div>
      )}
    </Fragment>
  );
};

export type CatalyticActivityData = {
  commentType: Comment;
  reaction?: {
    name: string;
    reactionReferences: { databaseType: string; id: string }[];
    ecNumber: string;
    evidences?: Evidence[];
  };
}[];

type CatalyticActivityProps = {
  comments?: CatalyticActivityData;
};

const CatalyticActivityView: React.FC<CatalyticActivityProps> = ({
  comments,
}) => {
  if (!comments || comments.length <= 0) {
    return null;
  }

  let rheaReactionCount = 0;
  return (
    <Fragment>
      <h4>Catalytic Activity</h4>
      {comments.map(catalyticActivity => {
        if (!catalyticActivity.reaction) {
          return null;
        }
        const rheaReactionReferences = catalyticActivity.reaction.reactionReferences.filter(
          isRheaReactReference
        );
        const rheaId =
          rheaReactionReferences.length > 0 &&
          getRheaId(rheaReactionReferences[0].id);
        rheaReactionCount += +!!rheaId;
        return (
          <span className="text-block" key={catalyticActivity.reaction.name}>
            <strong>{catalyticActivity.reaction.ecNumber}</strong>
            {` ${catalyticActivity.reaction.name}`}
            {catalyticActivity.reaction.evidences && (
              <UniProtEvidenceTag
                evidences={catalyticActivity.reaction.evidences}
              />
            )}
            {!!rheaId && (
              <RheaReactionVisualizer
                rheaId={rheaId}
                show={rheaReactionCount === 1}
              />
            )}
          </span>
        );
      })}
    </Fragment>
  );
};

export default CatalyticActivityView;
