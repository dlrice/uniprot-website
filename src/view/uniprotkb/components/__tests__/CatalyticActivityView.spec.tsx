import React from 'react';
import { render } from '@testing-library/react';
import CatalyticActivityView, {
  getRheaId,
  isRheaReactionReference,
  RheaReactionVisualizer,
} from '../CatalyticActivityView';
import CatalyticActivityUIDataJson from '../__mocks__/CatalyticActivityUIData.json';

describe('Catalytic activity', () => {
  test('should render catalytic activity', () => {
    const { asFragment } = render(
      <CatalyticActivityView comments={CatalyticActivityUIDataJson} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('RheaReactionVisualizer', () => {
  test('should render RheaReactionVisualizer', () => {
    const { asFragment } = render(
      <RheaReactionVisualizer rheaId={12345} show={false} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('getRheaId', () => {
  test('should return id 12345 from RHEA:12345', () => {
    expect(getRheaId('RHEA:12345')).toEqual(12345);
  });
  test('should return null from RHEA-COMP:12345', () => {
    expect(getRheaId('RHEA-COMP:12345')).toEqual(null);
  });
});

describe('isRheaReactReference', () => {
  test('should return true when database=Rhea and id string is RHEA:12345', () => {
    expect(
      isRheaReactionReference({ databaseType: 'Rhea', id: 'RHEA:12345' })
    ).toEqual(true);
  });
  test('should return true when database=ChEBI and id string is CHEBI:57287', () => {
    expect(
      isRheaReactionReference({ databaseType: 'ChEBI', id: 'CHEBI:57287' })
    ).toEqual(false);
  });
});
