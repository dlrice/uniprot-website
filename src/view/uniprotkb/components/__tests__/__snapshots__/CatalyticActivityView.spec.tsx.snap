// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalytic activity should render catalytic activity 1`] = `
<DocumentFragment>
  <h4>
    Catalytic Activity
  </h4>
  <p>
    <strong>
      EC:12112
    </strong>
     Reaction name
  </p>
</DocumentFragment>
`;
